src/libcore/intrinsics.rs

```diff
@@ -959,59 +959,131 @@ extern "rust-intrinsic" {
     /// value is not necessarily valid to be used to actually access memory.
     pub fn arith_offset<T>(dst: *const T, offset: isize) -> *const T;
 
-    /// Copies `count * size_of<T>` bytes from `src` to `dst`. The source
-    /// and destination may *not* overlap.
+    /// Copies `count * size_of::<T>()` bytes from `src` to `dst`. The source
+    /// and destination must *not* overlap.
     ///
-    /// `copy_nonoverlapping` is semantically equivalent to C's `memcpy`.
+    /// For regions of memory which might overlap, use [`copy`] instead.
+    ///
+    /// `copy_nonoverlapping` is semantically equivalent to C's [`memcpy`].
+    ///
+    /// [`copy`]: ./fn.copy.html
+    /// [`memcpy`]: https://www.gnu.org/software/libc/manual/html_node/Copying-Strings-and-Arrays.html#index-memcpy
     ///
     /// # Safety
     ///
-    /// Beyond requiring that the program must be allowed to access both regions
-    /// of memory, it is Undefined Behavior for source and destination to
-    /// overlap. Care must also be taken with the ownership of `src` and
-    /// `dst`. This method semantically moves the values of `src` into `dst`.
-    /// However it does not drop the contents of `dst`, or prevent the contents
-    /// of `src` from being dropped or used.
+    /// `copy_nonoverlapping` is unsafe because it dereferences a raw pointer.
+    /// The caller must ensure that `src` points to a valid sequence of type
+    /// `T`.
+    ///
+    /// # Undefined Behavior
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The region of memory which begins at `src` and has a length of
+    ///   `count * size_of::<T>()` bytes must be *both* valid and initialized.
```

Comment:
It is both unclear to me what "valid" means (a single live allocation..?). 

Also I don't understand why src must be initialized. Where did you get this from? This would make it impossible to implement realloc in a pure rust allocator (which must copy uninitialized memory, since it doesn't know what part of the buffer is initialized).

Reply:
"valid" was an attempt to communicate "allocated memory whose access doesn't violate LLVM's pointer aliasing rules" without overwhelming the reader. We might want to move this stuff out (see next comment).

My only insight was that, according to the [reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), undefined behavior includes "Reads of undef (uninitialized) memory".

I agree that my understanding was too strict, but I'm not sure what the proper constraint is.